### PR TITLE
How to Set up Rust API Servers with the Diseal ORM and the PostgreSQL

### LINKED ISSUE #5840: How to Set up Rust API Servers with the Diseal ORM and the PostgresSQL
#### Proposed title of article
Your title should be descriptive of the article/tutorial. Be Specific. 

Use keyword research to improve your article's SEO.

#### Proposed article introduction
*We're looking for the first 2-3 paragraphs of your article that appropriately summarize what your article will be about.*
We will use this writing snippet to help us assess the overall quality before approval so ensure it has been proofread, spellchecked and grammarchecked.

ORM (Object Relation Mapping) is a common technique for storing, retrieving, updating, and deleting data from a database specifically within a relational database. Common languages such as Node.js have different ORM libraries that help you achieve the general application of ORM when connecting the database and the application. The application wants to access the data from a database. However, it cannot do that directly. An ORM helps your application connect to a database. The big advantage is that an ORM helps you create data schemas and relations within your application. Whenever you need to change a specific database field, you only so it in your application with just a few lines of codes to change. This helps you avoid the hectic work of recreating your databases every time to match your new data registry.

The concept of ORM is widely supported by a number of languages such as Rust. Rust uses the Diseal framework to help you write your schema queries within your rust application. In this guide, we learn more about the Rust Diseal and finally set up a Rust API that uses Diseal to communicate with a PostgreSQL relational database.

#### Key takeaways 
What are the 3-5 most important things the reader should understand or be able to do after reading this article?
Use this area to get your ideas down on the bulk of your article or tutorial.

- What is an ORM
- What is Diseal
- Why the Diseal ORM
- Diesel migration
- Setup postgres
- Setup diesel
- Set up a Diseal Rust API server
- Write a new record to the PostgresSQL
- Show all the PostgresSQL records

#### Article quality
How is your article unique? Tell us what makes your approach different from similar articles that have published on the same topic?
Is yours more in-depth? Does it cover additional topics? Do you provide handy tips or anecdotal advice?

This tutorial will help the reader learn more about the ORM and understand how to use ORM in Rust using Diseal. The reader will also be able to create a Rust API server that uses Diseal to connect to PostgreSQL then create and retrieve the data that the application has stored in the database. This will help the reader understand the use of the concept of ORM using the Rust Diseal in an ideal application.

#### References
Please list links to any published content/research that you intend to use to support/guide this article.

### Conclusion
Finally, remove the Pre-Submission advice section and all our blockquoted notes as you fill in the form before you submit.
We look forwarding to reviewing your topic suggestion.

### Templates to use as guides
- [How To Guide Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/how-to-guide-template.md)
- [Software Review Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/software-review-template.md)
- [Tutorial Template](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/tutorial-template.md)

### PR DESCRIPTION
**Pull request by wanjcarol**
#### Formatting and structure checks
- [x] Have you used the correct folder and file structure? - **see [Contribution Guidelines](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/CONTRIBUTING.md)**
- [x] Is your article properly formatted in Markdown? - **see [Github Markdown guide](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet)**
- [x] Have you used our preferred heading structure? - **[H3 (###) and up](https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet#headers)**
- [x] Does your article follow any of the suggested structure formats? - **see [example formats](https://github.com/section-io/engineering-education/blob/master/CONTRIBUTING.md#example-format-structure)**

- [x] Is your article over 750 words? - **Tool**: [Wordcounter.net](https://wordcounter.net)
- [x] Is the article you are submitting an in-depth and unique article? Does it go beyond what is in the official docs and what is covered in other blog sites. See these articles as [examples](https://github.com/section-engineering-education/engineering-education/blob/master/new_contributors/resources-page.md).
- [x] Does your article provide enough value and detail about your topic? - **Articles should be clear, accurate, and fully explained.**
- [x] Can your article be understood by beginners? - **Assume the audience is smart but has no prior exposure to the common terminology in your article.**

- [x] Have you included a hero image that is **600x400 pixels**, under **300KB in size**, and have the copyright to use it? - **Tools:** [Pixlr Image Editor](https://pixlr.com/e) for resizing and [Unsplash](https://unsplash.com) for Creative Commons images
      
#### Grammar & plagiarism checks
- [x] Have you spellchecked and grammar checked your article? - **Tools:** [Write&Improve](https://writeandimprove.com/?lang=en-GB) - [ProWritingAid](https://prowritingaid.com/) - [Grammarly](https://grammarly.com) 
- [x] Please place your article through a 3rd party plagiarism checker. We suggest using Quetext, [this tool](https://www.quetext.com/) is free and has a daily limit. **We typically accept articles with 10% or less.**
- [x] Have you checked your article for readability? - **Tool:** [Hemmingway](http://www.hemingwayapp.com)
- [x] Have you added sources for quotes and images that aren't yours?

#### Technical checks
- [x] Are your code snippets properly formatted for syntax highlighting - **see [Syntax guide](https://gohugo.io/content-management/syntax-highlighting)**
- [x] Have you checked your code runs correctly and you've highlighted all necessary dependencies for installation?
- [x] Are the software programs and packages you're highlighting in your article up to date, using current versions and not deprecated?

This closes #5840

### CODE REVIEW

**Comment by jerimkaura** [comment]
Hello @wanjcarol 

Thank you for submitting this great article. I however feel you have not included the conclusion section that provides a summary of whatever is captured in the article.  Providing this recap not only adds value to the article and its format but also comes in line with the conventional article format by section.io. 
Screenshot from 2022-01-08 09-07-29

Take a look at this screenshot. Davis is providing a summary of his article which completes it well.

Let me know if you have any challenges with this.

**Comment by wanjcarol** [comment]
Yes @jerimkaura , am adding that right away

**Comment by WanjaMIKE** [comment]
Apart from the boiler plate code, could you @wanjcarol correct the other sentences showing plagiarism. Thanks

[New Scan 5_10 PM Copyleaks Report.pdf](https://github.com/section-engineering-education/engineering-education/files/7864822/New.Scan.5_10.PM.Copyleaks.Report.pdf)

**Comment by WanjaMIKE** [comment]
@jerimkaura could you check if this article is similar to https://www.freecodecamp.org/news/building-powerful-graphql-servers-with-rust/

**Comment by wanjcarol** [comment]
I have updated the sentences @WanjaMIKE

**Comment by jerimkaura** [comment]
> @jerimkaura could you check if this article is similar to https://www.freecodecamp.org/news/building-powerful-graphql-servers-with-rust/

Hello @WanjaMIKE , 

I did a check on the article. They are share implementation concepts but are completely different.

File: content/articles/how-to-set-up-rust-api-servers-with-the-diseal-orm-and-the-postgresql/index.md [hunk 1]
```diff
@@ -0,0 +1,452 @@
+ORM (Object Relation Mapping) is a common technique for storing, retrieving, updating, and deleting data from a database, specifically within a relational database. Common languages such as Node.js have different ORM libraries that help you achieve the general application of ORM when connecting the database and the application. The application wants to access the data from a database. However, it cannot do that directly. An ORM helps your application connect to a database. The significant advantage is that an ORM helps you create data schemas and relations within your application. Whenever you need to change a specific database field, you only so it in your application with just a few lines of codes to change. This helps you avoid the hectic work of recreating your databases every time to match your new data registry.
+
```
Comment (jerimkaura) [inline content/articles/how-to-set-up-rust-api-servers-with-the-diseal-orm-and-the-postgresql/index.md:2]: This paragraph is bulky, consider breaking into smaller understandable related sentences.

File: content/articles/how-to-set-up-rust-api-servers-with-the-diseal-orm-and-the-postgresql/index.md [hunk 2]
```diff
@@ -0,0 +1,452 @@
+ORM (Object Relation Mapping) is a common technique for storing, retrieving, updating, and deleting data from a database, specifically within a relational database. Common languages such as Node.js have different ORM libraries that help you achieve the general application of ORM when connecting the database and the application. The application wants to access the data from a database. However, it cannot do that directly. An ORM helps your application connect to a database. The significant advantage is that an ORM helps you create data schemas and relations within your application. Whenever you need to change a specific database field, you only so it in your application with just a few lines of codes to change. This helps you avoid the hectic work of recreating your databases every time to match your new data registry.
```
Comment (jerimkaura) [inline content/articles/how-to-set-up-rust-api-servers-with-the-diseal-orm-and-the-postgresql/index.md:1]: Is it lines of code or lines of codes?

File: content/articles/how-to-set-up-rust-api-servers-with-the-diseal-orm-and-the-postgresql/index.md [hunk 3]
```diff
@@ -0,0 +1,452 @@
...
+
+The concept of ORM is widely supported by many languages such as Rust. Rust uses the Diseal framework to help you write your schema queries within your rust application. In this guide, we learn more about the Rust Diseal and finally set up a Rust API that uses Diseal to communicate with a PostgreSQL relational database.
```
Comment (jerimkaura) [inline content/articles/how-to-set-up-rust-api-servers-with-the-diseal-orm-and-the-postgresql/index.md:3]: Many Languages should be more than one, not just just a single language as mentioned.

File: content/articles/how-to-set-up-rust-api-servers-with-the-diseal-orm-and-the-postgresql/index.md [hunk 4]
```diff
@@ -0,0 +1,452 @@
...
+
+This tutorial will help you learn more about the ORM and understand how to use ORM in Rust using Diseal. We will also create a Rust API server that uses Diseal to connect to PostgreSQL and then generate and retrieve the application's data stored in the database. This will help you understand the use of the concept of ORM using the Rust Diseal in an ideal application.
```
Comment (jerimkaura) [inline content/articles/how-to-set-up-rust-api-servers-with-the-diseal-orm-and-the-postgresql/index.md:5]: This statement is repeated. Summarize it together with line 5

File: content/articles/how-to-set-up-rust-api-servers-with-the-diseal-orm-and-the-postgresql/index.md [hunk 5]
```diff
@@ -0,0 +1,452 @@
...
+- [Setting up a Rust application](#setting-up-a-rust-application)
+- [Set up the GraphQL Schema](#set-up-the-graphql-schema)
+- [Set up the GraphQL server](#set-up-the-graphql-server)
+- [Setting up diesel](#setting-up-diesel)
+- [Handling queries](#handling-queries)
+- [Handling mutation](#handling-mutation)
+- [Conclusion](#conclusion)
+
+### Setting up a Rust application
+Rust uses Cargo to set up and run its applications. Cargo gets installed once the Rust compiler gets installed in your computer. To set up a basic Rust application, proceed to your desired project location, open the terminal, and run the following command to initialize the project using [cargo](https://doc.rust-lang.org/cargo/).
```
Comment (jerimkaura) [inline content/articles/how-to-set-up-rust-api-servers-with-the-diseal-orm-and-the-postgresql/index.md:26]: Explain to the user what cargo is.

**Review by jerimkaura — CHANGES_REQUESTED** [review]
Hello @wanjcarol 

Thanks for this article.:rocket:  I like how well you formatted the article :tada: However, there seem to be some readability issues with your article that are worth looking at. Not a problem :handshake: just countercheck keenly to address them well. I mentioned a few but not across the whole article.

Work on readability and clarity to speed up the review process.

**Comment by wanjcarol** [comment]
Check the changes @jerimkaura

File: content/articles/how-to-set-up-rust-api-servers-with-the-diseal-orm-and-the-postgresql/index.md [hunk 6]
```diff
@@ -0,0 +1,478 @@
...
+            .expect("Error saving new todo");
+
+        todo
+    }
+}
+```
+
+The `create_todo()` receives the new todo as a parameter, establishes database connection, inserts the todo, and returns the newly inserted todo.
+
+Replace `EmptyMutation` from a schema with `MutationRoot`:
```
Comment (jerimkaura) [inline content/articles/how-to-set-up-rust-api-servers-with-the-diseal-orm-and-the-postgresql/index.md:446]: Why are we conducting this step?

**Review by jerimkaura — CHANGES_REQUESTED** [review]
Hello @wanjcarol :wave: 

You have a great article with good content :rocket: However, there are minor enhancements that we need to do for the clarity of the article.

Remember, it's best practice to publish well-refined articles that readers find easy to read and comprehend :blush:

I recommend that you go over the suggestions I made in the article and work on them to enhance reliability, clarity and understanding of your code snippets.

Reach out to me via slack in case you need any assistance.

Thank you!

File: content/articles/how-to-set-up-rust-api-servers-with-the-diseal-orm-and-the-postgresql/index.md [hunk 7]
```diff
@@ -0,0 +1,478 @@
...
+Define a function to create the schema.
+
+```rust
+pub fn create_schema() -> Schema {
+    Schema::new(QueryRoot {}, EmptyMutation::new())
+}
+```
+
+### Set up the GraphQL server
+Now proceed to `main.rs` and set up the HTTP server to ensure that the schema is passed and called. However, first, update the imports as follows.
```
[the remark this answers is not in the document]
Comment (wanjcarol) [reply content/articles/how-to-set-up-rust-api-servers-with-the-diseal-orm-and-the-postgresql/index.md:156]: In the `src` folder, create a `graphql_schema.rs` file and import `EmptyMutation` and `RootNode` from `juniper` as shown below. Then implement GraphQL schema as shown in the following steps;

I had already stated this at the begging of this `### Set up the GraphQL schema`

**Comment by wanjcarol** [comment]
I have added the minor enhancements @jerimkaura

**Review by jerimkaura — COMMENTED** [review]
@wanjcarol 

Great work Proceeding to tech review.

Thanks.

**Review by jerimkaura — APPROVED** [review]
Hey @wanjcarol 

Great article here :tada:

Check out the changes I made to improve on future articles.

Thanks for the great work!

File: content/articles/how-to-set-up-rust-api-servers-with-the-diseal-orm-and-the-postgresql/index.md [hunk 8]
```diff
@@ -0,0 +1,506 @@
+---
+layout: engineering-education
+status: publish
+published: true
+url: /how-to-set-up-rust-api-servers-with-the-diseal-orm-and-the-postgresql/
```
Comment (lalith1403) [inline content/articles/how-to-set-up-rust-api-servers-with-the-diseal-orm-and-the-postgresql/index.md:5]: Shorten the URL. This is too long.

**Comment by jerimkaura** [comment]
Hey @wanjcarol 

I renamed your article folder to shorten the URL. Please upload the images to the new folder and delete the old folder. 

Let me know if you have any difficulties via slack. @jerimkaura

**Comment by wanjcarol** [comment]
@jerimkaura and @lalith1403 This should be good to go

**Review by jerimkaura — APPROVED** [review]
This can now be finalized @lalith1403

**Pull request merged**